### PR TITLE
Use jenkins user inside docker jobs instead of root

### PR DESCRIPTION
**Pull request by scpeters**
I have noticed some times when Jenkins fails to clear a workspace for a job. After investigating the machines over ssh, I have noticed the remaining folders contain files owned by `root`. This is caused by our use of the `root` user in the docker-based Ubuntu CI jobs. To avoid using the `root` user, this pull request creates a user in the Dockerfile that matches the current user, and switch to this user before using commands that touch the workspace folders.

I have tested with the following job:

* [Build Status](https://build.osrfoundation.org/job/sdformat-ci-main-bionic-i386/14/) https://build.osrfoundation.org/job/sdformat-ci-main-bionic-i386/14/

In order for this job to pass, I had to manually call `sudo chown -R jenkins:jenkins` on both the build machine's workspace folder and on the shared `/srv/ccache` folder. To merge this; I recommend taking all the Linux nodes of build.osrfoundation.org offline, then sanitizing the workspace and `/srv/ccache` folders of each Linux node, then bringing them back online.

**edit**
commands to run on migration:
```bash
sudo find /var/lib/jenkins/workspace -user root -exec chown jenkins {} \;
# chown symbolic links too
sudo find /var/lib/jenkins/workspace -user root -exec chown -h jenkins {} \;
sudo chown jenkins -R /srv/ccache
```

### CODE REVIEW

**Comment by j-rivero** [comment]
I think that the PR is a good step forward to reduce the use of root in the CI pipeline. It is probably the easiest one (more complicate would probably be the execution of docker without sudo) but this one is important.  Thank Steve.

Code looks good to me. I would like to check a couple of things more: GPU build with this branch and debbuilder with this branch.  My goal is to execute the merge and do the manual operations in the next two days.

**Comment by scpeters** [comment]
> I think that the PR is a good step forward to reduce the use of root in the CI pipeline. It is probably the easiest one (more complicate would probably be the execution of docker without sudo) but this one is important. Thank Steve.
> 
> Code looks good to me. I would like to check a couple of things more: GPU build with this branch and debbuilder with this branch. My goal is to execute the merge and do the manual operations in the next two days.

sounds good. It might be useful to schedule a meeting with you, me and the current build farmer @mjcarroll to coordinate the merge and manual operations

**Comment by j-rivero** [comment]
I made some changes to this PR:

 * Remove workarounds in place inside Jenkins DSL  61f11d6
 * Remove workarounds inside the docker_run script
 * Add some missing entries for sudo when using `make install` b864fef

**Comment by j-rivero** [comment]
> sounds good. It might be useful to schedule a meeting with you, me and the current build farmer @mjcarroll to coordinate the merge and manual operations

I was thinking on performing all the operations during Wednesday Europe morning if we have a clear plan with the steps ready.  I'm happy to take any help.

**Comment by scpeters** [comment]
have you run any test builds with these latest changes?

**Comment by j-rivero** [comment]
Made some modifications to this PR (sorry I should have used a different branch) and run a battery of tests:

 * debbuilder job: [Build Status](https://build.osrfoundation.org/job/_test_nonroot_ign-msgs8-debbuilder/6/)
 * ci job with GPU: [Build Status](https://build.osrfoundation.org/job/_test_nonroot_ignition_rendering-ci-ign-rendering5-bionic-amd64/7/)
 * abichecker job: [Build Status](https://build.osrfoundation.org/job/_test_nonroot_sdformat-abichecker-any_to_any-ubuntu_auto-amd64/5/)
 * install job: [Build Status](https://build.osrfoundation.org/job/_test_nonroot_sdformat-install-sdformat10_pkg-bionic-amd64/1/)
 * ros2 ci job: [Build Status](https://build.osrfoundation.org/job/_test_nonroot_ros2_gazebo_pkgs-ci-default_foxy-focal-amd64/2/)
 * gazebo job: [Build Status](https://build.osrfoundation.org/job/_test_nonroot_gazebo-ci-gazebo11-bionic-amd64-gpu-nvidia/1/)

I think we are ready to re-review @scpeters

File: jenkins-scripts/docker/lib/_install_nvidia_docker.sh [hunk 1]
```diff
@@ -2,12 +2,12 @@ INSTALL_NVIDIA_DOCKER1="""
...
-apt-get install -y docker-ce
+sudo apt-get update
+sudo apt-get install -y docker-ce
 echo '# END SECTION'
 
 echo '# BEGIN SECTION: install nvidia-docker1 (in docker)'
-apt-get install -y wget nvidia-340 nvidia-modprobe
+sudo apt-get install -y wget nvidia-340 nvidia-modprobe
 wget -P /tmp https://github.com/NVIDIA/nvidia-docker/releases/download/v1.0.1/nvidia-docker_1.0.1-1_amd64.deb
 dpkg -i /tmp/nvidia-docker*.deb && rm /tmp/nvidia-docker*.deb
```
Comment (scpeters) [inline jenkins-scripts/docker/lib/_install_nvidia_docker.sh:12]: I expect this line and the `apt-key` and `add-apt-repository` commands may need `sudo` as well

**Comment by scpeters** [comment]
thanks for the additional testing. I think we can merge soon and catch further issues when they arise

File: jenkins-scripts/docker/lib/_install_nvidia_docker.sh (hunk 1 above)
Comment (j-rivero) [reply jenkins-scripts/docker/lib/_install_nvidia_docker.sh]: you are right 079a093

**Comment by j-rivero** [comment]
> thanks for the additional testing. I think we can merge soon and catch further issues when they arise

File: jenkins-scripts/docker/lib/docker_generate_dockerfile.bash [hunk 2]
```diff
@@ -417,12 +411,30 @@ echo '# END SECTION'
 BUILDSH_CCACHE
 fi
 
+cat >> Dockerfile << DELIM_DOCKER_USER
+# Create a user with passwordless sudo
+ARG USERID
+ARG USER
+RUN adduser --uid \$USERID --gecos "Developer" --disabled-password \$USER
```
Comment (tfoote) [inline jenkins-scripts/docker/lib/docker_generate_dockerfile.bash:418]: I would recommend a slightly more complete arguments to the user. In particular set the group ID as well. 

I have a good example at: https://github.com/osrf/rocker/blob/acbbff59d6aef6a9654eabc0ff7e3f823936d6e3/src/rocker/templates/user_snippet.Dockerfile.em#L9

Comment (j-rivero) [reply jenkins-scripts/docker/lib/docker_generate_dockerfile.bash]: Thanks! I've included GID handlng in  ef59662

**Comment by j-rivero** [comment]
Another short round of tests:

 * ign-rendering [Build Status](https://build.osrfoundation.org/job/_test_nonroot_ignition_rendering-ci-ign-rendering5-bionic-amd64/9/)
 * ign-msgs8-debbuilder: [Build Status](https://build.osrfoundation.org/job/_test_nonroot_ign-msgs8-debbuilder/7/)

**Comment by j-rivero** [comment]
Changed machines:
 * drogon
 * maria
 * optimus
 * r2d2
 
Merging-

**Pull request merged**